[vcpkg registries] Registries: Take 2 (#13590)
microsoft/vcpkg: docs/specifications/registries-2.md

is not permitted. Future extensions may include things like remote archives or git repositories,
or may allow `"registry-path"` in git registries.

Note that the `"versions"` field should _always_ be additive;
deleting existing entries is unsupported and may result in bad behavior.
The only modification to existing entries that is allowable is moving the backing store
for the build files, assuming that the new build files are equivalent to the old build files.
(For example, a filesystem registry might have a new way of laying out where ports are)

### Example of a registry database entry file

Note: This file assumes that the versions RFC has been implemented,
and thus that minimum versions are required;
the syntax may change in the time between now and finishing the implementation.

This example is of `ogre`, since this port has both features and dependencies;
remember that this file would be `😇/o-/ogre.json`.

```json
{
  "baseline-version": { "version-string": "1.12.7" },
  "versions": [
    {
      "version-string": "1.12.7",
      "git-tree": "466e96fd2e17dd2453aa31dc0bc61bdcf53e7f61",
      "dependencies": [

>         { "name": "freeimage", "version-string": "3.18.0", "port-version": 13 },

Reviewer: Shouldn't these be version requirements instead of version declarations? 

Since we are adding version information here as a way to impose version requirements on packages that have existed before this was possible.

```json
[{ "name": "freeimage", "version-string": "3.18.0", "port-version": 13 }, ...]
```

Author: You're right.

this means that the ports directory is no longer anything but a data store.
This also means that the existing rules around locations of ports is no longer required;
however, it will still keep getting followed for the main repository,
and it's recommended that other registries follow the same pattern to make contributing easier.

An important consideration is making it easy for people to contribute new ports,
and new versions of ports; thus, all of the common registry updates can be done via vcpkg commands.
The most common command will be `vcpkg add-port`, which will update the registry database
with the new port or version that the user has added.
We will also add checking to the PR testing to make certain that the registry database is up-to-date
with the ports directory.
More information on the `vcpkg add-port` command is in the
[Adding Ports to the Registry Database](#adding-ports-to-the-registry-database) section.

## What does the registry database look like?

The folder structure is up for bikeshedding; however right now the design of it is as follows.
We don't wish to have the same problem as we do right now,

> where there are nearly 1500 entries in a single directory.

Author: Yes absolutely. One advantage of this new registry structure means we can modify the ports tree without breaking anything (assuming people update to the new version of vcpkg).

We solve this by placing each database entry into `😇/<first character of port name>-/<port name>.json`.

> For example, the database entry for 7zip shall be at `😇/7-/7zip.json`.

Reviewer 2: ```suggestion
For example, the database entry for 7zip is `😇/7-/7zip.json`.
```

there will be a `"registries"` property that is an object.
This object will contain a `"git"` field, which is an array of git-registry objects,
that contain:

* `"repository"`: The `"repository"` field from the git registry object
* `"commit"`: The commit ID that HEAD referred to at the time.

For example, a `vcpkg-lock.json` might look like:

```json
{
  "registries": {
    "git": [
      {
        "repository": "https://github.com/microsoft/vcpkg",
        "commit": "6185aa76504a5025f36754324abf307cc776f3da"
      }
    ]
  }
}
```

#### `vcpkg update`

You'll notice that once the repository is added the first time,
there is only one way to update the repository to the tag at a later date - deleting the lock file.
We additionally want to add support for the user updating the registry by themselves -
they will be able to do this via the `vcpkg update` command.
The `vcpkg update` command will, for each git registry,
update the registry and repoint the `"commit"` field in `vcpkg-lock.json` to the latest `HEAD`.

There is no way to update only one git registry to a later date, since versions are strictly additive.

## Git Registries: Implementation on Disk

> There are two actual implementations on disk to consider here: the implementation of the registry database,

Reviewer 2: ```suggestion
There are two implementations on disk to consider here: the implementation of the registry database,
```

and once we have the database entries for the ports, accessing the port data from the git tree object.

> Both of these implementations shall be placed in the vcpkg cache home (shared by binary caching archives).

Reviewer 2: ```suggestion
Both of these implementations are placed in the vcpkg cache home (shared by binary caching archives).
```

On unix, this is located at `$XDG_CACHE_HOME/vcpkg` if the environment variable exists,
otherwise `$HOME/.cache/vcpkg`; on Windows, it's located at `%LOCALAPPDATA%\vcpkg`.
We will add a new folder in this cache, called `registries`, which will contain all the data we need.

First, we'll discuss the registry database.

### Registry Database

> At `registries/git`,

Reviewer 2: ```suggestion
At `$CACHE_ROOT/registries/git`,
```

?

Reviewer 2: (And similar for all paths that need to be relative to that; since it's confusing whether we are talking about the cache location vs. a vcpkg tree from there)

> we'll create a new git repository root which shall contain all information from all git registries,

Reviewer 2: ```suggestion
we'll create a new git repository root which contains all information from all git registries,
```

since the hashes should be unique, and this allows for deduplication
across repositories which have the same commits (e.g., for mirrors).
In order to get the data from git registries, we simply `fetch` the URL of the git registry.

In order to grab a specific database entry from a git registry, `git show` is used to grab the
file from the right commit: `git show <commit id> -- 😇/<first character>-/<portname>.json`.

One unfortunate thing about having one directory being used for all vcpkg instances on a machine is

> that it's very possible to have an issue with concurrency - for example, after `fetch`ing the latest HEAD

Reviewer 2: ```suggestion
that it's possible to have an issue with concurrency - for example, after `fetch`ing the latest HEAD
```

each reference has a set of versions that are tested against each other,
and this is a major feature of vcpkg.
We wish to have the same feature in the new versioning world,
and so we'll have a set of baseline versions in the registry database.

There are two kinds of baselines: the `"default"` baseline,
which is used by default, and also named baselines.
Baseline names must be identifiers, in the same vein as ports.
Additionally, for git registries,
if the user uses a baseline that does not have an entry,
then vcpkg attempts to use the `"default"` entry at the reference specified by the baseline.

These baselines are placed in `port_versions/baseline.json`.
This is an object mapping baseline names to baseline objects,
where baseline objects map port names to version objects.
A version object contains exactly one version field (one of `"version-string"`, `"version"`, etc.),
and optionally `"port-version"`.

[git object ID]: https://git-scm.com/book/en/v2/Git-Internals-Git-Objects

### Example of a baseline file

The following is a reasonable baseline.json for a filesystem registry that only has two ports:

```json
{
  "default": {
    "abseil": { "version-string": "2020-03-03" },

>     "zlib": { "version-string": "1.2.11", "port-version": 9 }

Reviewer: I think we should make this a generic version field that does not imply any kind of scheme. 

I would suggest `"version"` but it is kind of unfortunate that it overlaps with the "Relaxed" scheme.
```
"abseil": { "version": "2020-03-03" },
"zlib": { "version": "1.2.11", "port-version": 9 }
```

Another suggestion would be "version-text"
```
"abseil": { "version-text": "2020-03-03" }
"zlib": { "version-text": "1.2.11", "port-version": 9 }
```
Or we could change the spec for versioning so that `"version"` does not imply relaxed and instead make it explicit `"version-relaxed"`.

  },
  "old": {
    "abseil": { "version-string": "2019-02-11" },
    "zlib": { "version-string": "1.2.11", "port-version": 3 }
  },
  "really-old": {
    "zlib": { "version-string": "1.2.9" }
  }
}
```

### Example of a registry database entry file

Note: This file assumes that the versions RFC has been implemented,
and thus that minimum versions are required;
the syntax may change in the time between now and finishing the implementation.

This example is of `ogre`, since this port has both features and dependencies;
remember that this file would be `port_versions/o-/ogre.json`.

```json
[
  {
    "version-string": "1.12.7",

>     "git-tree": "466e96fd2e17dd2453aa31dc0bc61bdcf53e7f61",

Reviewer: Same as with baselines use a generic version field.

```
{
  "version": "1.12.7",
  "git-tree": "466e96fd2e17dd2453aa31dc0bc61bdcf53e7f61"
}
```

Author: This definitely shouldn't use a generic field; there's no other place to grab the version scheme from.

it's simply a way to get the files for a port.
As an example, let's assume that the registry is laid out something like this:

```
<registry>/
  😇/
    baseline.json
    a-/
      abseil.json
      asmjit.json
    o-/
      ogre.json
  ports/
    a-/
      abseil/
        2020-03-03_7/
          vcpkg.json
          portfile.cmake
          ...
        2020-03-03_8/
          vcpkg.json
          portfile.cmake
          ...
        ...
      asmjit/
        2020-05-08/
          CONTROL
          portfile.cmake
          ...
        2020-07-22/
          vcpkg.json
          portfile.cmake
          ...
    o-/
      ogre/
        1.12.7/
          ...
        1.12.1/
          ...
    ...
  ...
```

Then, let's look at updating `asmjit` to latest.

The current manifest file, in `asmjit/2020-07-22/vcpkg.json` looks like:

```json
{
  "name": "asmjit",
  "version-string": "2020-07-22",
  "description": "Complete x86/x64 JIT and Remote Assembler for C++",
  "homepage": "https://github.com/asmjit/asmjit",
  "supports": "!arm"
}
```

while the current `😇/a-/asmjit.json` looks like:

```json
[
  {
    "version-string": "2020-07-22",

>     "registry-path": "/ports/a-/asmjit/2020-07-22"

Reviewer 3: This looks like magic without knowing the implementation in detail.
Something like `${base}/ports/a-/asmjit/2020-07-22` is in my eyes clearer.
(or I am just VS Code trained where some variables are also available in settings.json)

  },
  {
    "version-string": "2020-05-08",
    "registry-path": "/ports/a-/asmjit/2020-05-08"
  }
]
```

with `😇/baseline.json` looking like:

```json
{
  "default": {
    ...,
    "asmjit": { "version-string": "2020-07-22" },
    ...
  }
}
```

and we'd like to update to 2020-10-08.
We should first copy the existing implementation to a new folder:

```sh

> $ cp -r ports/a-/asmjit/2020-07-22 ports/a-/asmjit/2020-10-08

Reviewer 4: Do you think it would be a good idea to add an option to the `vcpkg add-port` command so that this part is done "automatically" for us? Something like

1. `vcpkg add-port ./ports/a-/asmjit/2020-10-08 --base ports/a-/asmjit/2020-07-22 --port-type registry-path`
1. Make the edits required to `ports/a-/asmjit/2020-10-08` to update to latest.

This would remove a manual manipulation, and it prevents repeating `ports/a-/asmjit/2020-10-08` twice, removing at the same time an opportunity to make a typo. :slightly_smiling_face:

remember that this file would be `😇/o-/ogre.json`.

```json
[
  {
    "version-string": "1.12.7",
    "git-tree": "466e96fd2e17dd2453aa31dc0bc61bdcf53e7f61",
  },
  {
    "version-string": "1.12.1",
    "port-version": 1,
    "git-tree": "0de81b4f7e0ec24966e929c2ea64e16c15e71d5e",
  },
  ...

> ]

Reviewer: Database generator implemented according to spec in PR #13777
